crates/http/src/spin.rs

```diff
@@ -1,231 +1,241 @@
-        // object as opposed to headers.
-        for (keys, val) in crate::compute_default_headers(req.uri(), raw, base, host)? {
-            res.push((Self::prepare_header_key(keys[0]), val));
-        }
-
-        Ok(res)
-    }
-
-    fn prepare_header_key(key: &str) -> String {
-        key.replace('_', "-").to_ascii_lowercase()
-    }
-
-    fn append_headers(res: &mut http::HeaderMap, src: Option<Vec<(String, String)>>) -> Result<()> {
-        if let Some(src) = src {
-            for (k, v) in src.iter() {
-                res.insert(
-                    http::header::HeaderName::from_str(k)?,
-                    http::header::HeaderValue::from_str(v)?,
-                );
-            }
-        };
-
-        Ok(())
-    }
-
-    fn params(uri: &Uri) -> Result<Vec<(String, String)>> {
-        match uri.query() {
-            Some(q) => Ok(url::form_urlencoded::parse(q.as_bytes())
-                .into_owned()
-                .collect::<Vec<_>>()),
-            None => Ok(vec![]),
-        }
-    }
-}
-
-fn contextualise_err(e: anyhow::Error) -> anyhow::Error {
-    if e.to_string()
-        .contains("failed to find function export `canonical_abi_free`")
-    {
-        e.context(
-            "component is not compatible with Spin executor - should this use the Wagi executor?",
-        )
-    } else {
-        e
-    }
-}
-
-#[cfg(test)]
-mod tests {
-    use super::*;
-
-    #[test]
-    fn test_spin_header_keys() {
-        assert_eq!(
-            SpinHttpExecutor::prepare_header_key("SPIN_FULL_URL"),
-            "spin-full-url".to_string()
-        );
-        assert_eq!(
-            SpinHttpExecutor::prepare_header_key("SPIN_PATH_INFO"),
-            "spin-path-info".to_string()
-        );
-        assert_eq!(
-            SpinHttpExecutor::prepare_header_key("SPIN_RAW_COMPONENT_ROUTE"),
-            "spin-raw-component-route".to_string()
-        );
-    }
-}
+use crate::{
+    spin_http::{Method, SpinHttp},
+    ExecutionContext, HttpExecutor, RuntimeContext,
+};
+use anyhow::Result;
+use async_trait::async_trait;
+use http::Uri;
+use hyper::{Body, Request, Response};
+use spin_engine::io::{ModuleIoRedirects, ModuleIoRedirectsTypes};
+use std::{net::SocketAddr, str, str::FromStr};
+use tokio::task::spawn_blocking;
+use tracing::log;
+use wasmtime::{Instance, Store};
+
+#[derive(Clone)]
+pub struct SpinHttpExecutor;
+
+#[async_trait]
+impl HttpExecutor for SpinHttpExecutor {
+    async fn execute(
+        &self,
+        engine: &ExecutionContext,
+        component: &str,
+        base: &str,
+        raw_route: &str,
+        req: Request<Body>,
+        _client_addr: SocketAddr,
+        follow: bool,
+    ) -> Result<Response<Body>> {
+        log::trace!(
+            "Executing request using the Spin executor for component {}",
+            component
+        );
+
+        let mut mior = if let ModuleIoRedirectsTypes::FromFiles(clp) = engine.config.module_io_redirects.clone() {

```

Comment:
It would be super useful to have an integration test for this - I know it's fiddly because CLP can only be injected through hosting, but it would really help with confidence when changing things in the I/O subsystem.

Reply:
The way I've tested this feature locally was:
- get an app w/ `spin_loader::from_file`,
- create `ExecutionContextConfiguration` object w/ `CustomLogPipes`,
- get a builder w/ `spin_engine::Builder::with_engine`,
- `HttpTrigger::configure_execution_context`,
- `builder.build()`
- get `component_triggers` from `app`,
- `HttpTrigger::new` w/ all that,
- `.run()` the `http_trigger`, and then
- `curl` it to see the result of the `.wasm` module's `println!`s.

Are you suggesting recreating something like that in the integration tests? If not, how would you go about it?

Reply:
Something like that would be great.  Unfortunately the integration tests only have infrastructure for the normal Spin binary at the moment (via `SpinTestController`), so there's new infra to be built here.  If you don't want to tackle that, raise an issue and I'll get onto it once this is merged.  (And if you could share your existing manual test code in the issue, that would be a big timesaver!)

Reply:
Hmmm~ I just don't know how much time I can afford to allocate there. SOOO... If that's ok then, I think I'll pass that onto you and help w/ reviewing.

Here's where we are using the `CustomLogPipies` like I mentioned: https://github.com/Mossaka/containerd-wasm-shims/blob/main/containerd-shim-spin-v1/src/main.rs

Reply:
Fair enough - thanks for the pointer!